src/app/RequestLogger.ts

```diff
@@ -0,0 +1,137 @@
+import { APIGatewayProxyEventV2, APIGatewayProxyStructuredResultV2, CloudFrontHeaders, CloudFrontRequestEvent, CloudFrontRequestResult, CloudFrontResultResponse } from "aws-lambda";
+import { cutTooLongString } from "../utils/helpers";
+
+export type LogPackage = {
+    trace: {
+        id: string;
+        amazonTraceId?: string;
+        userAgent: string;
+        sourceIp?: string;
+    };
+    request: {
+        method: string;
+        path: string;
+        query: string;
+    };
+    response: {
+        statusCode: number;
+        contentLength: number;
+        body?: string; // For error diagnostics
+    };
+};
+
+export type RequestLoggerSettings = {
+    disable?: {
+        sourceIp?: boolean;
+    }
+};
+
+export default class RequestLogger {
+    private readonly request: APIGatewayProxyEventV2 | CloudFrontRequestEvent;
+    private readonly configuration: RequestLoggerSettings;
+
+    constructor(request: APIGatewayProxyEventV2 | CloudFrontRequestEvent, configuration?: RequestLoggerSettings) {
+        this.request = request;
+        this.configuration = configuration || {};
+    }
+
+    public log(response: APIGatewayProxyStructuredResultV2 | CloudFrontRequestResult): void {
+        const log = this.parseLogPackage(this.request, response);
+
+        if (this.configuration.disable?.sourceIp) {
+            log.trace.sourceIp = undefined;
+        }
+
+        if (typeof log.response.statusCode !== "number" || log.response.statusCode >= 400) {
+            log.response.body = cutTooLongString(log.response.body, 1000); // Limit logged body length to 1000 characters
+            console.error(JSON.stringify(log));
+        } else {
+            delete log.response.body; // Remove body from successful requests logs
+            console.log(JSON.stringify(log));
+        }
+    }
+
+    private parseLogPackage(request: APIGatewayProxyEventV2 | CloudFrontRequestEvent, response: APIGatewayProxyStructuredResultV2 | CloudFrontRequestResult): LogPackage {
+        // identify if event is APIGatewayProxyEventV2 or CloudFrontRequestEvent
+        if (typeof (request as any).requestContext !== "undefined") {
+            return this.parseApiGatewayEvent(request as APIGatewayProxyEventV2, response as APIGatewayProxyStructuredResultV2);
+        }
+        return this.parseCloudFrontEvent(request as CloudFrontRequestEvent, response as CloudFrontRequestResult);
+    }
```

Comment:
Ite monesti kipuilee tälläsissä caseissa että parametri on joko tyyppiä A, B (tai jopa C) ja sen mukaan toimenpiteet. Monesti miettii et oisko fiksuinta antaa joku ylimääräinen parametri joka kertoo asian (joka tietysti toimii kun on 2 vaihtoehtoa, kuten tässä), tai sit oisko kliinimpää käyttää jotain type guardeja asian tsekkaamiseen. Toki samahan tässäkin tehdään tavallaan, tosin one linerina ja ajaa kyllä asian kun ei tätä tsekkiä tarvii tehdä kun tässä kohtaa ainoastaan. https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

Tuli mieleen et oisko sitä valibottia tms voinut hyödyntää ns. runtime tsekkailuun skeemoineen, tosin sillä kai pitäis sitä parse funkkaria käyttää eli se menis sellaseen try catch kikkailemiseen..

Reply:
Tämä kipuilu on tuttua kaikissa dynaamisesti tyypitetyissä kielissä ja typescriptin kanssa melkeinpä se kliini ratkaisu löytyy helpoiten käyttelemällä niitä apukirjastoja. Hauskasti samaa käsikirjan kohtaa linkkailtiin kommenteissa :) 

Toi valibot olis mukana tässä repossa eri haarassa: https://github.com/Virtual-Finland-Development/codesets/pull/31